### ISSUE TITLE
Definition of terms related to uncertainty

### ISSUE THREAD

**Issue by betatim** [MechCoder/scikit-optimize#23]
Some thoughts on "uncertainty". This issue was inspired by @MechCoder's comment in #9. The first part of this issue tries to correctly define various terms that often get used interchangeably and are easy to confuse (I confidently predict that I will make at least one error in this post). Once we have defined the terms, we can decide which of them we need in order to evaluate various acquisition functions.

[Standard deviation](http://mathworld.wolfram.com/StandardDeviation.html) (\sigma): this is the square root of the variance. Can be calculated for any sample no matter what distribution the samples come from.

[Standard error](http://mathworld.wolfram.com/StandardError.html) (of the mean): `\sigma / \sqrt(N)` a measure of the uncertainty associated with the estimated value of the mean.

[Confidence interval](http://mathworld.wolfram.com/ConfidenceInterval.html) (CI): The N% confidence interval will contain the measured value N% of the time. Alice wants to estimate the value of a parameter `t`, so she constructs an estimator `that` as well as a CI. The 68% CI (around `that`) will contain the true value `t` in 68% of experiments (that is we clone Alice and repeat what she did many times).

N% quantile: The N% quantile starts at negative infinity and goes until a point `x`, think of it as the integral of the p.d.f. between `-inf` and `x` which equals N%.

If `that` is distributed according to a normal distribution then the 68% CI is [`that` - `sigma`, `that` + `sigma`].

For a normal distribution `mu`-`sigma` = the 16% quantile.

For our purposes we have a surrogate model (a GP or what have you) for the true, expensive function `f`. At a given point `x` our best estimate of the true value of `f` is the mean `mu(x)` of our surrogate model.

-------------

Now my understanding runs out -> need help.

What is the band we get from a GP and then feed into EI and friends? Is it the "standard error on the mean" or "68% confidence interval" or "68% [credible interval](https://en.wikipedia.org/wiki/Credible_interval)" or something else?

**Comment by glouppe** [comment]
> What is the band we get from a GP and then feed into EI and friends?

A GP defines a probability over functions that could have generated the observed data (x_i, y_i). At a fixed x, you therefore also get a conditional distribution for y|x (by integrating over all possible functions and evaluating their value at x). This conditional distribution is a model of the true distribution of y|x, if you could observe infinitely many values y for the same x (e.g. because the relation between x and y involves some random fluctuations). The band you get from a GP corresponds to the standard deviation of that  conditional distribution, under the model assumed by the GP. 

To summarize, what is important for us is to use models that not only define some value of y given x, but a whole distribution of y given x. And at the bare minimum, we require to at least be able to compute the mean and standard deviation of that modeled distribution. 

---

Note that I never use the term "uncertainty" - we should ban that word, or then be very explicit about the uncertainty of what quantity we are referring to. Also, standard errors and confidence intervals are concepts we should not mistake, as they are a tied to the estimation of a statistic. For example, we are not interested about the uncertainty of some estimate of the mean value of y|x, because we could be very precisely estimate the mean value of y|x, despite the fact that y|x has high variance (in such case, the uncertainty of the estimate of the mean would be low).

**Comment by MechCoder** [comment]
Thanks for the long note. It's a bit confusing to me as well, but here are my thoughts.

>  Standard deviation, can be calculated for any sample no matter what distribution the samples come from.

The standard deviation that we calculate using the formula described here [Standard deviation](https://en.wikipedia.org/wiki/Standard_deviation#Unbiased_sample_standard_deviation) is an unbiased estimate of the actual standard deviation of the actual distribution. If we need to calculate the exact standard deviation, then we need to go by `sqrt(E(x**2) - E(x)**2)` which needs the distribution of x.

In a Gaussian Process, we are able to get the exact distribution of the target variable that we want to predict given the training data, which is a gaussian. Because of this we are able to get the actual standard deviation (and not an estimate).

**Comment by MechCoder** [comment]
@glouppe I think I've asked this to you before, but how does knowing the y|x at different quantiles, help us calculate the standard deviation (so as to be fed into an acquistion function), without having any explicit assumptions about what distribution y|x follows?

**Comment by glouppe** [comment]
>  without having any explicit assumptions about what distribution y|x follows?

Oh but you do make assumptions, which may be wrong but this is fine. E.g., using a GP with some specified kernel is equivalent to making regularity assumptions about the true underlying function, and therefore also about the shape of the true distribution y|x. A GP is model, not the true thing.

**Comment by glouppe** [comment]
> how does knowing the y|x at different quantiles, help us calculate the standard deviation 

Well for example the most common practice is to decide to use some quantile (e.g. the 68th) as an approximation mean+sigma under our model (which is itself an approximation). This is a choice that we make.

**Comment by MechCoder** [comment]
Ok, by assumptions I did not mean the bias of the model as explained by your first comment. In the sense, once we fix the model to be a GP , the conditional distribution can be analytically derived. However, it seemed that even after fixing our model to a GBRT, still no assumptions can be made about the conditional.

Your second comment however answers my question. It should be pretty straightforward to try.

**Comment by glouppe** [comment]
Yeah, or with a forest-based model, you can also use the empirical mean and the empirical variance of the training samples in the leaf reached by x.

**Comment by betatim** [comment]
A notebook with some old exploration I did on [modelling "uncertainty"](https://gist.github.com/betatim/f3949b7d8d04d6c25f01). Here `truth` is the expensive function we are trying to model and the noise represents the noise inherent to observing that expensive function. In this particular setup the noise level varies as a function of the parameter, but you can switch to constant noise. I built these to convince myself that you can model this kind of noise with different kinds of models (not just GPs). I couldn't justify the theory behind why this should work, I came to this via "intuition".

--------

Is what we really want mean+sigma? I thought what people really wanted is the CI. You can use standard deviation to get the +sigma in the case of a gaussian distribution, but CI came first then mean+sigma instead of the other way around.

**Comment by MechCoder** [comment]
Well, it depends on the acquisition function. You can see the derivation here for GP's here on page 2 (http://www.cse.wustl.edu/~garnett/cse515t/files/lecture_notes/12.pdf) . It should reduce to something similar given you are modelling the conditional as a Gaussian.

It is also important not to get confused between CI and the std. I think if the distribution is gaussian, then there is a nice relation between both, but the CI models the uncertainty in the mean due to difference in sampling and the std is just inherent to the distribution.

I get similar confusions in mind, but a good majority of them are during exams.

**Comment by MechCoder** [comment]
@glouppe 

> Yeah, or with a forest-based model, you can also use the empirical mean and the empirical variance of the training samples in the leaves reached by x.

If I am right, this will require modifying the DecisionTree code to return the std as well?

**Comment by glouppe** [comment]
Not convinced we need to do that now. To build a working proof of concept, all these empirical statistics can be computed using the `apply` method in trees.

**Comment by glouppe** [comment]
Might be nice to write a notebook about these concepts, that could be part of the documentation. What do you think?
